### PR TITLE
change sequence_bias type of SequenceBiasLogitsProcessor to list, add…

### PR DESCRIPTION
**Pull request by VladOS95-cyber**
# What does this PR do?
Introduce new sequence_bias parameter type as List[List[List[int], float]] for SequenceBiasLogitsProcessor and config tests for all logits processors

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/32416
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante @zucchini-nlp @ArthurZucker 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by VladOS95-cyber** [comment]
Hi @gante, please, review this PR. I completely messed with previous one https://github.com/huggingface/transformers/pull/33362. This is cleaner and included config tests for almost all processors

**Comment by VladOS95-cyber** [comment]
Just add @zucchini-nlp and @ArthurZucker in case if @gante is busy

File: tests/generation/test_configuration_utils.py [hunk 1]
```diff
@@ -224,6 +256,411 @@ def test_generation_mode(self):
         config = GenerationConfig()
         self.assertEqual(config.get_generation_mode(assistant_model="foo"), GenerationMode.ASSISTED_GENERATION)
 
+    def test_serialize_generation_sequence_bias(self):
```
Comment (zucchini-nlp) [inline tests/generation/test_configuration_utils.py:259]: I love these tests ❤️

and I think @gante will review the PR himself soon

Comment (VladOS95-cyber) [reply tests/generation/test_configuration_utils.py:259]: Ok, got it, thanks!

**Comment by VladOS95-cyber** [comment]
@zucchini-nlp Hi! Do I have to fix failed tests somehow? Because it looks like it is not related to my changes at all

**Comment by zucchini-nlp** [comment]
Failing tests are okey, they are also failing in other PRs. You can try rebasing main to see if they are already fixed, the last time it was still broken in one of my PRs

**Comment by VladOS95-cyber** [comment]
> Failing tests are okey, they are also failing in other PRs. You can try rebasing main to see if they are already fixed, the last time it was still broken in one of my PRs

Ok, got it, thanks a lot!

**Comment by VladOS95-cyber** [comment]
Hi @gante! Just a kindly reminder ping in order to not loose this PR!

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1064,8 +1064,8 @@ class SequenceBiasLogitsProcessor(LogitsProcessor):
     </Tip>
 
     Args:
-        sequence_bias (`Dict[Tuple[int], float]`):
-            Dictionary that maps a sequence of tokens to its bias term. Positive biases increase the odds of the
+        sequence_bias (`List[List[Union[List[int], float]]]`):
+            List of lists that maps a sequence of tokens to its bias term. Positive biases increase the odds of the
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1068]:
```suggestion
            List of lists that maps a sequence of tokens to its bias term (e.g. `[[[10, 45], -2.0], [[64], -7.5]]`). Positive biases increase the odds of the
```

Since the format is more complex, let's add a concrete example right at the start :) That way the format should be immediately obvious

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -1191,9 +1196,31 @@ def _validate_arguments(self):
...
                 f"{sequence_bias}."
             )
-        if any(not isinstance(bias, float) for bias in sequence_bias.values()):
+
+        def all_token_bias_pairs_are_valid(sequence):
+            return (
+                isinstance(token_bias_pair, list)
+                and all(isinstance(token_id, (int, np.integer)) and token_id > 0 for token_id in token_bias_pair)
+                or isinstance(token_bias_pair, float)
+                for token_bias_pair in sequence
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1205]: This function is called when `isinstance(sequence_bias, list)`. However, if I'm not mistaken, this condition only applied to the old `dict` format -- if so, let's remove these lines

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1191,9 +1196,31 @@ def _validate_arguments(self):
...
+        ):
+            raise ValueError(
+                f"Each element in `sequence_bias` has to be a non-empty list of lists of positive integers and float, but is "
+                f"{sequence_bias}."
+            )
+        if isinstance(sequence_bias, dict) and any(not isinstance(bias, float) for bias in sequence_bias.values()):
             raise ValueError(f"`sequence_bias` has to be a dict with floats as values, but is {sequence_bias}.")
 
+    def _convert_list_arguments_into_dict(self):
+        """BC: we used to accept `dict{tuple of tokens: float}`"""
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1219]:
```suggestion
        """BC: we used to accept `dict{tuple of tokens: float}` directly, now we expect a list"""
```
(more precise)

File: tests/generation/test_configuration_utils.py [hunk 5]
```diff
@@ -224,6 +256,411 @@ def test_generation_mode(self):
...
+        generation_config = GenerationConfig()
+        sequence_bias = [[[45, 67], -0.6], [[89], 1.2]]
+        generation_config.sequence_bias = sequence_bias
+        with tempfile.TemporaryDirectory("test-generation-config") as tmp_dir:
+            generation_config.save_pretrained(tmp_dir)
+            new_config = GenerationConfig.from_pretrained(tmp_dir)
+        self.assertSequenceEqual(new_config.sequence_bias, sequence_bias)
+
+        bias_logits_processor = SequenceBiasLogitsProcessor(new_config.sequence_bias)
+        self.assertIsNotNone(bias_logits_processor.sequence_bias)
```
Comment (gante) [inline tests/generation/test_configuration_utils.py:270]: Can we check instead that `bias_logits_processor.sequence_bias` is the same as `sequence_bias`? It is more informative than `None` :)

(this pattern should be applied on all new tests)

File: tests/generation/test_configuration_utils.py (hunk 1 above)
Comment (gante) [inline tests/generation/test_configuration_utils.py:259]:
```suggestion
class GenerationConfigSerializationTest(unittest.TestCase):
    def test_serialize_generation_sequence_bias(self):
```
These tests deserve their own category 💛

**Review by gante — APPROVED** [review]
Thank you for iterating @VladOS95-cyber -- and especially for adding the new tests. I really appreciate the work 🙏 

IMO the PR could be merged as is. I've added a few improvement suggestions for long-term maintenance. 

Regarding red CI -- try rebasing the PR. If it's still red after that, I will dive to figure out what's wrong :)

**Comment by gante** [comment]
@LysandreJik some context :D

this PR: 
1. Changes the type of the `sequence_bias`. The previous type was not json serializable. We keep BC, but all documentation points towards the serializable type (so that users don't hit the limitation)
2. @VladOS95-cyber kindly added tests for serialization of logits processors arguments in the generation config 💛

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (VladOS95-cyber) [reply src/transformers/generation/logits_process.py:1068]: sure, done

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (VladOS95-cyber) [reply src/transformers/generation/logits_process.py:1205]: it is actually a wrong logic here, I just noticed, I changed it to verify that first item in the list is exactly list of positive integers and the second one is a float

File: src/transformers/generation/logits_process.py (hunk 4 above)
Comment (VladOS95-cyber) [reply src/transformers/generation/logits_process.py:1219]: done

File: tests/generation/test_configuration_utils.py (hunk 1 above)
Comment (VladOS95-cyber) [reply tests/generation/test_configuration_utils.py:259]: agree, done

File: tests/generation/test_configuration_utils.py (hunk 5 above)
Comment (VladOS95-cyber) [reply tests/generation/test_configuration_utils.py:270]: done

**Comment by VladOS95-cyber** [comment]
> Thank you for iterating @VladOS95-cyber -- and especially for adding the new tests. I really appreciate the work 🙏
> 
> IMO the PR could be merged as is. I've added a few improvement suggestions for long-term maintenance.
> 
> Regarding red CI -- try rebasing the PR. If it's still red after that, I will dive to figure out what's wrong :)

Hi, I just pushed changes based on your comments and did rebase, still the same test issues

**Comment by gante** [comment]
@VladOS95-cyber I've checked internally -- it is a known issue and it is being fixed :) 

By the time @LysandreJik approves the PR it should be fixed

**Review by LysandreJik — APPROVED** [review]
That's a very impressive additional test suite!

The change looks good to me, thanks for updating the docs alongside it.

**Comment by VladOS95-cyber** [comment]
> That's a very impressive additional test suite!
> 
> The change looks good to me, thanks for updating the docs alongside it.

Hi @LysandreJik sure, no problem. Should I do smth else, rebase with commit to fix failed tests or anything?

**Comment by gante** [comment]
Hey @VladOS95-cyber -- rebasing should get us rid of the red CI!

**Comment by VladOS95-cyber** [comment]
> Hey @VladOS95-cyber -- rebasing should get us rid of the red CI!

Hey @gante, ok, no worries, I'll do that

**Comment by VladOS95-cyber** [comment]
> green CI, perfect!
> 
> @VladOS95-cyber thank you for iterating with us and, more importantly, for making `transformers` better for everyone 💛

@gante Thank you too! Let's keep working on it!
